### PR TITLE
Some preliminary work towards making trans "collector driven".

### PR DESCRIPTION
**Pull request by michaelwoerister**
The `trans::collector` already collects all translation items and `trans::partitioning` distributes these translation items into codegen units. The changes in this PR provide the following extensions to this functionality:
1. Drop-glue is handled more accurately now, knowing about the difference between `DropGlueKind::Ty` and `DropGlueKind::TyContents`.
2. The partitioning module now supports the `FixedUnitCount` strategy which more or less corresponds to the partitioning one gets via supplying `-Ccodegen-units` today.
3. The partitioning scheme also takes care of assigned LLVM declarations to codegen units, not just definitions (declarations for external items not yet implemented).

It's debatable whether declarations should be handled by the partitioning scheme or whether they should just be emitted on demand.

### CODE REVIEW

**Comment by michaelwoerister** [comment]
cc @nikomatsakis

File: src/librustc_data_structures/bitvec.rs [hunk 1]
```diff
@@ -52,7 +52,7 @@ impl BitVector {
 
     pub fn grow(&mut self, num_bits: usize) {
         let num_words = u64s(num_bits);
-        let extra_words = self.data.len() - num_words;
+        let extra_words = (num_words as i64) - (self.data.len() as i64);
```
Comment (nagisa) [inline src/librustc_data_structures/bitvec.rs]: `if self.data.len() > num_words { self.data.resize(num_words, 0) }` instead?

Comment (michaelwoerister) [inline src/librustc_data_structures/bitvec.rs]: Yeah, that would be nicer. I'll add the test too.

**Comment by michaelwoerister** [comment]
Changed the implementation of `BitVector::grow` to @nagisa's nicer version and made the corresponding test case more comprehensive.

File: src/test/codegen-units/partitioning/local-generic.rs [hunk 2]
```diff
@@ -9,7 +9,7 @@
 // except according to those terms.
 
 // ignore-tidy-linelength
-// compile-flags:-Zprint-trans-items=eager
+// compile-flags:-Zprint-trans-items=eager -Zincremental=""
```
Comment (nikomatsakis) [inline src/test/codegen-units/partitioning/local-generic.rs]: why `-Z incremental=""`?

File: src/librustc_trans/collector.rs [hunk 3]
```diff
@@ -252,12 +254,74 @@ impl<'tcx> Hash for TransItem<'tcx> {
     }
 }
 
-pub type InliningMap<'tcx> = FnvHashMap<TransItem<'tcx>, FnvHashSet<TransItem<'tcx>>>;
+/// Maps every translation item to all translation items it references in its
+/// body.
+pub struct ReferenceMap<'tcx> {
+    index: FnvHashMap<TransItem<'tcx>, (usize, usize)>,
```
Comment (nikomatsakis) [inline src/librustc_trans/collector.rs]: it'd be great to have a comment on what these two usizes are -- a range?

File: src/librustc_trans/collector.rs [hunk 4]
```diff
@@ -252,12 +254,74 @@ impl<'tcx> Hash for TransItem<'tcx> {
...
+        self.index.insert(source, (start_index, end_index));
+    }
+
+    // Internally iterate over all items referenced by `source` which will be
+    // made available for inlining.
+    pub fn with_inlining_candidates<F>(&self, source: TransItem<'tcx>, mut f: F)
+        where F: FnMut(TransItem<'tcx>)
+    {
+        if let Some(&(start_index, end_index)) = self.index.get(&source)
+        {
```
Comment (nikomatsakis) [inline src/librustc_trans/collector.rs]: Nit: move `{` to previous line

**Comment by michaelwoerister** [comment]
I'll take the failed test as an opportunity to address the nits.

**Pull request merged**